lib/dotdiff/element_handler.rb

```diff
@@ -9,27 +9,29 @@ def initialize(driver, elements = DotDiff.xpath_elements_to_hide)
 
     def hide
       elements.each do |xpath|
-        if element_exists?(xpath)
-          driver.execute_script("#{js_element(xpath)}.style.visibility = 'hidden'")
-        end
+        driver.execute_script(script(xpath, 'hidden')) if elements_exists?(xpath)
       end
     end
 
     def show
       elements.each do |xpath|
-        if element_exists?(xpath)
-          driver.execute_script("#{js_element(xpath)}.style.visibility = ''")
-        end
+        driver.execute_script(script(xpath, '')) if elements_exists?(xpath)
       end
     end
 
-    def js_element(xpath)
-      "document.evaluate(\"#{xpath}\", "\
-      "document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue"
+    def script(xpath, visibility)
+      # this is done like so instead of a single pass over all elements due to a bug in Firefox:
+      # https://greasyfork.org/en/forum/discussion/12223/xpath-iteratenext-fails-in-firefox
+      xpath += visibility == 'hidden' ?
+        "[not(contains(@style, 'visibility'))]" :
+        "[contains(@style, 'visibility: hidden')]"
+      "var elem; while (elem = document.evaluate(\"#{xpath}\", document, "\
+        'null, XPathResult.ORDERED_NODE_ITERATOR_TYPE, null).iterateNext()) '\
+        "{ elem.style.visibility = '#{visibility}'; }"
     end
 
-    def element_exists?(xpath)
-      driver.find(:xpath, xpath, wait: DotDiff.max_wait_time, visible: :all) rescue nil
+    def elements_exists?(xpath)
+      driver.all(:xpath, xpath, wait: DotDiff.max_wait_time, visible: :all).any?
```

Comment:
This will only wait up until it finds one matching element - what happens if you have more than 1 element that you need to hide that doesn't display until later, but is already required to be hidden?

Reply:
Actually you're right, this code is not needed at all. I believe `dotdiff` should not wait for anything. The contract with the user is that he/she should already have awaited everything that *must* be there prior to taking the screenshot. This makes perfect sense and is the easiest to reason about for both the user and the library. This way `dotdiff` doesn't actually need to wait for anything and just hides/shows whatever is there. Please tell me if you disagree but I am pretty confident that this is the right way.